### ISSUE TITLE
Tracking Issue for Contracts

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#128044]
This is a tracking issue for the MCP "Contracts" (https://github.com/rust-lang/compiler-team/issues/759).
The feature gate for the issue is `#![feature(rustc_contracts)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.
Discussion comments will get marked as off-topic or deleted.
Repeated discussions on the tracking issue may lead to the tracking issue getting locked.

### Steps

- [ ] Implement the MCP (cc @rust-lang/compiler)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Formatting for new syntax has been added to the [Style Guide] ([nightly-style-procedure])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs
[nightly-style-procedure]: https://github.com/rust-lang/style-team/blob/master/nightly-style-procedure.md 
[Style Guide]: https://github.com/rust-lang/rust/tree/master/src/doc/style-guide

### Unresolved Questions

Syntax bikesheds galore: I am not thrilled by the prospect of writing #[rustc_contracts::requires(for safety: ...)] all over the place.
For example, an earlier draft of this MCP had used #[safety::requires(...)], which I find far more palatable, but may pose serious hazards for compatibility with existing code bases.
I am leaving identification of "a good syntax" as an open problem.

Static vs dynamic semantics: My idealized contract system would allow contracts to inform both static verification and dynamic validation tools. Some contracts like forall! do not have an "obviously right" dynamic interpretation.

Safety post-obligations: The safety criteria for some unsafe methods is stated as a constraint on how the caller uses the return value from a method.
For example: str::as_bytes_mut says "The caller must ensure that the content of the slice is valid UTF-8 before the borrow ends and the underlying str is used."
This cannot be expressed as a mere safety::requires form as envisioned above.
(My best guess is that such contracts require a richer structure, such as an entiretly hypothetical #[safety::at_lifetime_end(|output| str::from_utf8(output).is_ok())], which could only be checked in the future, right before the &mut u8 borrow expires and the reference is again capable of being treated as a &mut str.)

Correctness invariants: As mentioned above, there is probably utility in being able to attach invariants to a type that are used for proving functional correctness. But it is not as clear where to establish the points where correctness invariants must be checked. It may make more sense here to use something like refinement types, where explicit method calls (potentially in ghost code) would (re)establish such invariants.

Purity: Do contracts need to be pure (i.e. have no non-local side-effects)? Some verification tools will require this.

Panic: How should a panicking expression within a contract be treated? Some of the above code was written with slice's get method rather than the index operator, in order to sidestep this question. But rather than disallowing panic entirely (i.e. treating it as a improperly written contract), it might be more natural to interpret a panic within a contract expression as a contract failure.

**Comment by traviscross** [comment]
We've approved this as as lang experiment:

- https://github.com/rust-lang/rust/pull/128045#issuecomment-2622237287

**Comment by codeOverFlow** [comment]
Hello, maybe a newbie question, but why use attributes instead of language syntax similar to `where` clause, and similar to what Ada does ?

**Comment by traviscross** [comment]
@codeOverFlow: We don't really use tracking issues for such questions and discussion.  It'd be better to raise your question over on Zulip or as a separate GH issue.

**Comment by RalfJung** [comment]
@celinval what's the status of the design document that you started here? Now that there is an implementation in rustc, it'd be good to have e.g. a page on the dev guide explaining the overall design, and in particular explaining the lowering of the attributes, as that can be quite hard to follow otherwise.

**Comment by celinval** [comment]
I just finished updating [the design document](https://hackmd.io/@celinaval/B1UEl3tUyl), and I included the changes to support adding contract to constant functions.I believe there's a request to hold a [design review](https://github.com/rust-lang/compiler-team/issues/813) with the compiler team, but I don't know when it will be.

I can create a PR to the dev-guide. Should I wait for the design review meeting?

**Comment by RalfJung** [comment]
Maybe it makes sense to wait until after the meeting then.

**Comment by Connor-GH** [comment]
> Static vs dynamic semantics: The idealized contract system would allow contracts to inform both static verification and dynamic validation tools. What's the best way to handle conditions that cannot be checked with both semantics.

There is prior art in static verification with the likes of [SPARK](https://en.wikipedia.org/wiki/SPARK_(programming_language)) or [flux-rs](https://github.com/flux-rs/flux).

This would probably be a feature best kept behind a feature gate due to the compilation cost you would have to pay for an SMT solver. An SMT solver would probably be outside of Rust's scope so consider that the alternative is using yet another third-party library that we have no control over (liquid-fixpoint). Additionally, the audience for this feature is almost exclusively the academic or safety-critical crowd, and the latter is unlikely to get along with this feature in Rust since they like ground-up proofs (think seL4). The former, however, is where Rust gets quite a few of its ideas from, so I see as a fit, albeit with a lot of work to do. Projects like Redox OS and the Linux kernel could make use of this and provide logical guarantees, not just memory safety guarantees.

**Comment by Nadrieril** [comment]
And as per [this comment](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/Lang.20experiment.3A.20contracts/near/558326389) I am now the t-lang liaison for this experiment! Moreover as per [this discussion](https://rust-lang.zulipchat.com/#narrow/channel/233931-t-compiler.2Fmajor-changes/topic/Contracts.3A.20primitive.20ownership.20assertions.E2.80.A6.20compiler-team.23942/near/554802689), I proposed @dawidl022 as owner of this experiment as they've been doing a lot of the implementation work.

If you have any questions, ping me on Zulip or come chat on the [dedicated stream](https://rust-lang.zulipchat.com/#narrow/channel/544080-t-lang.2Fcontracts).
